tests/Fixtures/TestBundle/Serializer/Denormalizer/RecoverPasswordDenormalizer.php

```diff
@@ -0,0 +1,54 @@
+<?php
+
+/*
+ * This file is part of the API Platform project.
+ *
+ * (c) Kévin Dunglas <dunglas@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace ApiPlatform\Core\Tests\Fixtures\TestBundle\Serializer\Denormalizer;
+
+use ApiPlatform\Core\Serializer\AbstractItemNormalizer;
+use ApiPlatform\Core\Tests\Fixtures\TestBundle\Dto\RecoverPasswordInput;
+use Symfony\Component\Serializer\Normalizer\ContextAwareDenormalizerInterface;
+use Symfony\Component\Serializer\SerializerAwareInterface;
+use Symfony\Component\Serializer\SerializerAwareTrait;
+
+class RecoverPasswordDenormalizer implements ContextAwareDenormalizerInterface, SerializerAwareInterface
+{
+    use SerializerAwareTrait;
+
+    /**
+     * {@inheritdoc}
+     */
+    public function denormalize($data, $class, $format = null, array $context = [])
+    {
+        // Use a new context to avoid using this same denormalizer twice
+        $newContext = $context;
+        unset($newContext['input_class']);
+
+        // The RecoverPasswordInput has a user relation that gets denormalized
+        $inputDto = $this->serializer->denormalize($data, $context['input_class'], $format, $newContext);
+        $inputDto->user->getEmail();
+
+        // Because we're in a PUT operation, we will use the retrieved object...
+        $resourceObject = $context[AbstractItemNormalizer::OBJECT_TO_POPULATE] ?? new $class();
+        // ...where we remove the credentials
+        $resourceObject->eraseCredentials();
+
+        return $resourceObject;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function supportsDenormalization($data, $type, $format = null, array $context = []): bool
+    {
+        return RecoverPasswordInput::class === ($context['input_class'] ?? null);
+    }
+}
```

Comment:
This example will be used in the documentation. @ragboyjr maybe that you have remarks about this?

Reply:
Is part of the new requirements of input/output classes are that we need to create our own normalizers/denormalizers?

Reply:
Yes, I'm working on a graph to explain how things works but it's definitely better then using persisters and providers.

Reply:
Interesting, I can't speak for everyone, but my use case with input/output classes would be that I'm creating a custom handler for a specific endpoint. So instead of allowing ApiP to use their defaults, I'm intercepting it.

With messenger enabled, this is incredibly elegant, you set your input class, it gets routed to messenger which then routes to a service tagged as a message handler which is responsible for handling the input dto, creating the resource, persist it, then return it.

Without messenger, I feel like the next best solution is you'd set the input class, create a custom controller for the specific operation, and set api_persist=false since your controller will handle it.

With these new changes, that flow gets a lot more complicated since we'd know need to setup a new normalizer/denormalizer to bring our array data into the input dto, and then route it into the system. I'm not sure business logic should be put into data persisters/providers or serializers.

My goal would be that input/output classes get automatically normalized/denormalized according to the custom services created via messenger or via a custom controller.

Reply:
So my understanding with api_persist=false was that it disables the write listener. Why would input_class=false disable that? Also `input_class=null` which is the default means use resource_class for input, `input_class=SomeClass` means use that for input, `input_class=false` means disable the write listener?

Reply:
It's a shortcut introduced by @dunglas in https://github.com/api-platform/core/commit/e10d658cb25c00b0f7172b5f9c74b47993ec181c#diff-0170cc702a42034d94c5d83ccadad1deR81. It totally makes sense when you say "output_class=false" it means that you don't want any output, the response will therefore be empty. With `input_class=false` it says "don't read my data" (skips the deserialization).

Reply:
Gotcha, that makes sense, I think you meant `output_class=class` is the same as `api_persist=false` not `input_class=false`.

Reply:
Yes sorry!